File: example/android/ElmyExample/app/src/main/java/dev/elmy/Elmy.kt
package dev.elmy

import android.app.Activity
import android.util.Log
import androidx.compose.Composable
import androidx.compose.Model
import androidx.ui.core.Text
import androidx.ui.layout.Column
import androidx.ui.layout.Row
import androidx.ui.material.Button
import androidx.ui.toHexString
import org.liquidplayer.javascript.JSContext
import org.liquidplayer.javascript.JSFunction

class ElmyApp(activity: Activity, source: String) {
    val state = ElmyState.empty()
    private val jsContext = JSContext()

    init {
        jsContext.setExceptionHandler { Log.e("Elmy", it.stack()) }
        val tick: JSFunction = object : JSFunction(jsContext, "elmyTick") {
            fun elmyTick(bytes: Array<Int>): Unit {
                activity.runOnUiThread {
                    state.update(bytes)
                }
            }
        }
        jsContext.property("elmyTick", tick)
        jsContext.evaluateScript(source)
    }

    fun message(msg: StateByte) : Unit {
        jsContext.evaluateScript("ElmyApp.ports.msg.send($msg)")
    }
}

@Composable
fun Elmy(app: ElmyApp) {
    Column {
        ElmyElement(app.state) { app.message(it) }
    }
}

@Composable
fun ElmyElement(bin: ElmyState, sendMsg: (msg: StateByte) -> Unit) {
    if (!bin.hasNext()) {
        return
    }

    when (val element = bin.uInt8()) {
        0xA0 -> ElmyContainer(bin, sendMsg)
        0xB0 -> Text(bin.string(bin.uInt32()))
        0xB3 -> {
            bin.skip(2u) // Skip Button attrs for now

            val msg = bin.uInt16()

            Button ({
                sendMsg(msg)
            }) {
                ElmyElement(bin, sendMsg)
            }
        }
        else -> Text("Unknown element ${element.toInt().toHexString()}")
    }
}

@Composable
fun ElmyContainer(bin: ElmyState, sendMsg: (msg: StateByte) -> Unit) {
    val container = bin.uInt8()

    bin.skip(2u)  // Skip Attributes for now

    @Composable
    fun children(bin: ElmyState) {
        val len = bin.uInt16()
        for (i in 0 until len) {
            ElmyElement(bin, sendMsg)
        }
    }

    when (container) {
        0x00 -> Row {
            children(bin)
        }
        0x02 -> Column {
            children(bin)
        }
        else -> Text("Unknown container ${container.toInt().toHexString()}")
    }
}

typealias StateByte = Int
typealias StateUpdate = Array<StateByte>

@Model
class ElmyState(private var bytes: Iterator<StateByte>) {
    companion object {
        fun empty(): ElmyState {
           return ElmyState(IntArray(0).iterator());
        }
    }

    fun update(bytes: StateUpdate) {
        this.bytes = bytes.iterator()
    }

    fun hasNext(): Boolean {
        return bytes.hasNext()
    }

    fun uInt8(): Int {
        return bytes.next()
    }

    fun uInt16(): Int {
        val left = uInt8()
        val right = uInt8()
        return left shl 8 or right
    }

    fun uInt32(): Int {
        return uInt16() shl 16 or uInt16()
    }

    fun string(len: Int): String {
        var string = ""
        for (i in 0 until len) {

            val byte = uInt8()

            when {
                byte in 0..128 ->
                    string += byte.toChar()
                //                (byte and 0xE0 == 0xC0) ->
                //                    string += (byte and 0x1F) shl 6 or
            }
        }

        return string
    }

    fun skip(count: UInt) {
        for (i in 0u until count) {
            bytes.next()
        }
    }
};

File: swift-ui/Sources/Elmy/Elmy.swift
//
//  Elmy.swift
//  ElmyExample
//
//  Created by Agustín Zubiaga on 2/17/20.
//  Copyright © 2020 Agustín Zubiaga. All rights reserved.
//

import SwiftUI
import BinaryKit
import JavaScriptCore

@available(OSX 10.15, *)
final class ElmyApp : ObservableObject {
    @Published var state: [UInt8] = []
    
    let jsContext = JSContext()
    
    init(_ source: String) {
        jsContext?.exceptionHandler = {
            (ctx: JSContext!, value: JSValue!) in
            print(value ?? "Unknown error")
        }

        let consoleLog: @convention(block) (String) -> Void = { message in
            print("Elm > " + message)
        }
        
        jsContext?.setObject(consoleLog, forKeyedSubscript: "_consoleLog" as NSCopying & NSObjectProtocol)

        jsContext?.evaluateScript("var console = { log: _consoleLog, warn: _consoleLog, error: _consoleLog }")

        TimerJS.registerInto(jsContext: jsContext!)
        
        let tick: @convention(block) (JSValue) -> Void = {value in
             if let bytes = value.toArray() as? [UInt8] {
                DispatchQueue.main.async {
                    self.state = bytes
                }
            }
        }
        
        jsContext?.setObject(tick, forKeyedSubscript: "elmyTick" as NSCopying & NSObjectProtocol)

        jsContext?.evaluateScript(source)
    }
    
    func message(msg: UInt16) {
        jsContext?.evaluateScript("ElmyApp.ports.msg.send(\(msg))")
    }
}


@available(OSX 10.15, *)
struct Elmy : View {
    @ObservedObject private var app: ElmyApp;
    
    init (_ source: String) {
        app = ElmyApp(source);
    }
    
    var body: some View {
        var bin = Binary(bytes: self.app.state)
        
        return elmyElement(&bin, message: self.app.message)
    }
}


@available(OSX 10.15, *)
func elmyElement(_ bin: inout Binary, message: @escaping (UInt16) -> Void) -> some View {
    do {
        switch (try bin.readByte()) {
        case 0xA0: // Container
            return AnyView(elmyContainer(&bin, message: message))
        case 0xB0: // Text
            return AnyView(Text(try decodeString(&bin)))
        case 0xB3: // Button
            _ = try bin.readBytes(quantitiy: 2); // Skip attrs
            
            let msg = UInt16(bigEndian: Data(try bin.readBytes(quantitiy: 2)).withUnsafeBytes { $0.pointee })
                        
            return AnyView(Button(action: {
                message(msg)
            }) {
                elmyElement(&bin, message: message)
            })
        case let b:
            let x = String(format:"%02X", b)
            return AnyView(Text("Unknown element: \(x)"))
        }
    } catch {
        return AnyView(Text("Error!"))
    }
}

@available(OSX 10.15, *)
func elmyContainer(_ bin: inout Binary, message: @escaping (UInt16) -> Void) -> some View {
    do {
        let container = try bin.readByte();
        
        try bin.readBytes(quantitiy: 2);
        
        let children = try decodeList(decode: { bin in elmyElement(&bin, message: message)}, bin: &bin)
        
        let content = ForEach(children, id: \.self.key, content: { $0.value })
        
        switch (container) {
        case 0x00: // Row
            return AnyView(HStack { content })
        case 0x02: // Column
            return AnyView(VStack { content })
        case let b:
            let x = String(format:"%02X", b)
            return AnyView  (Text("Unknown container: \(x)"))
        }
    } catch {
        return AnyView(Text("Error: readContainer!"))
    }
}

typealias Decoder<T> = (_ bin: inout Binary) -> T;

struct Keyed<T> {
    let key: Int
    let value: T
}

func decodeList<T>(decode: Decoder<T>, bin: inout Binary) throws -> Array<Keyed<T>> {
    let data = Data(try bin.readBytes(quantitiy: 2))
    let size = Int(UInt16(bigEndian: data.withUnsafeBytes { $0.pointee }))

    var array: Array<Keyed<T>> = []

    for i in 0..<size {
        array.append(Keyed(key: i, value: decode(&bin)))
    }

    return array
}

func decodeString(_ bin: inout Binary) throws -> String {
    let data = Data(try bin.readBytes(quantitiy: 4))
    let size = UInt32(bigEndian: data.withUnsafeBytes { $0.pointee })
    return try bin.readString(quantitiyOfBytes: Int(size))
}

// JavaScript Timers

let timerJSSharedInstance = TimerJS()

@objc protocol TimerJSExport : JSExport {

    func setTimeout(_ callback : JSValue,_ ms : Double) -> String

    func clearTimeout(_ identifier: String)

    func setInterval(_ callback : JSValue,_ ms : Double) -> String

}

// Custom class must inherit from `NSObject`
@objc class TimerJS: NSObject, TimerJSExport {
    var timers = [String: Timer]()

    static func registerInto(jsContext: JSContext, forKeyedSubscript: String = "timerJS") {
        jsContext.setObject(timerJSSharedInstance,
                            forKeyedSubscript: forKeyedSubscript as (NSCopying & NSObjectProtocol))
        jsContext.evaluateScript(
            "function setTimeout(callback, ms) {" +
            "    return timerJS.setTimeout(callback, ms)" +
            "}" +
            "function clearTimeout(indentifier) {" +
            "    timerJS.clearTimeout(indentifier)" +
            "}" +
            "function setInterval(callback, ms) {" +
            "    return timerJS.setInterval(callback, ms)" +
            "}"
        )
    }

    func clearTimeout(_ identifier: String) {
        let timer = timers.removeValue(forKey: identifier)

        timer?.invalidate()
    }


    func setInterval(_ callback: JSValue,_ ms: Double) -> String {
        return createTimer(callback: callback, ms: ms, repeats: true)
    }

    func setTimeout(_ callback: JSValue, _ ms: Double) -> String {
        return createTimer(callback: callback, ms: ms , repeats: false)
    }

    func createTimer(callback: JSValue, ms: Double, repeats : Bool) -> String {
        let timeInterval  = ms/1000.0

        let uuid = NSUUID().uuidString

        // make sure that we are queueing it all in the same executable queue...
        // JS calls are getting lost if the queue is not specified... that's what we believe... ;)
        DispatchQueue.main.async(execute: {
            let timer = Timer.scheduledTimer(timeInterval: timeInterval,
                                             target: self,
                                             selector: #selector(self.callJsCallback),
                                             userInfo: callback,
                                             repeats: repeats)
            self.timers[uuid] = timer
        })


        return uuid
    }

    @objc func callJsCallback(_ timer: Timer) {
        let callback = (timer.userInfo as! JSValue)

        callback.call(withArguments: nil)
    }
}
